### ISSUE TITLE
GitPackIndex cannot read 'slice' of null

### ISSUE THREAD

**Issue by johnwiseheart** [isomorphic-git/isomorphic-git#1163]
Using node:

```
const fs = require("fs")
const git = require("isomorphic-git")

git.log({ fs, dir: "/my-project", depth: 1, ref: "<sha>" }).then(res => console.log(res))
```

errors with:

```
UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'slice' of null
    at GitPackIndex.readSlice (/my-project/node_modules/isomorphic-git/index.cjs:2666:35)
```

This appears to be coming from: https://github.com/isomorphic-git/isomorphic-git/blob/main/src/models/GitPackIndex.js#L297 

I was not on the latest, but I tried upgrading and it's still happening.

What is strange is that its pretty intermittent - its only happening in my ci, and I can't seem to repro locally. Any idea what might be causing this? Perhaps some issue with the git repo? I tried a `git fsck` already but it didn't show up any issues.

**Comment by johnwiseheart** [comment]
It seems specific to the commit, as with an empty commit applied on top of the broken one, the command passes with no issues

**Comment by kdumontnu** [comment]
@johnwiseheart did you end up resolving this? I'm seeing a similar issue.

**Comment by Jazcash** [comment]
Getting this too, after a clone. If I clone with `noCheckout: true` then restart my app and do a `.checkout` then it works fine, but if I clone and checkout in the same instance of my app then I get the error

**Comment by jcubic** [comment]
Without a reproducible example, we can't do anything about it. But note that the original author doesn't work much on this project anymore. I'm a maintainer but I also don't write code myself. So if any of you want to debug this and find the root cause it would be great. And if anyone wants to fix the issue I will be able to merge it.

**Comment by Jazcash** [comment]
Changing the read function in FileSystem.js to log errors before returning null makes the problem more clear:

```
  async read(filepath, options = {}) {
    try {
      let buffer = await this._readFile(filepath, options)
      // Convert plain ArrayBuffers to Buffers
      if (typeof buffer !== 'string') {
        buffer = Buffer.from(buffer)
      }
      return buffer
    } catch (err) {
      if (err) {
        console.error(err);
        return null;
      }
    }
  }
```

`RangeError: Array buffer allocation failed`

My packfile here is 1.8GB so I'm pretty sure this comes down to a memory limitation.

**Comment by ikeyan** [comment]
In my case the error was `RangeError [ERR_FS_FILE_TOO_LARGE]: File size (4373791561) is greater than 2 GB`

**Comment by loynoir** [comment]
3 days ago, I benchmark `isomorphic-git` with ~2.5G repo yarnpkg/berry with 2.2M file, the test pass.

Now, I  rebuild devcontainer, or said system updated, I encounter this bug.

Maybe this is related to

- node version

- system

https://github.com/loynoir/git-bench/blob/733d6b7c7b3bae9f5655396b023e2889b302a4ae/tool.mjs#L36-L45

https://github.com/loynoir/git-bench/blob/733d6b7c7b3bae9f5655396b023e2889b302a4ae/tool.mjs#L103

**Comment by loynoir** [comment]
Seems [axkibe](https://github.com/axkibe) figure out what is going on.

Out of my interest, just FYI. @jcubic 

https://github.com/nodegit/nodegit/issues/1977#issuecomment-1500963241

> Out of some reason, isogit is not looking for the commit object in the pack file, this "fixes" it as workaround in berry:
>
> mv .git/objects/pack/pack-5594355c4b021332ab59e14ea79d5c46bf10ec68.pack .
> git unpack-objects < pack-5594355c4b021332ab59e14ea79d5c46bf10ec68.pack

**Comment by mojavelinux** [comment]
The real issue is that there is an IO error, but isomorphic-git is suppressing it. The problem starts at this line:

```js
const raw = (await this.pack).slice(start);
```

At this point, `this.pack` is a promise to read the file. If there's an IO error, the fs implementation returns null (see https://github.com/isomorphic-git/isomorphic-git/blob/162827ae4261495a56ef8598e3153868016296c2/src/models/FileSystem.js#L109). At this point, the call to `slice` fails. So we really don't know what the real error is.

For anyone who wants to find out, remove the catch in the read method and let's see what's going on.

**Comment by jcubic** [comment]
@mojavelinux thanks for the hints. So the problem this code:

```javascript
    if (!this.pack) {
      throw new InternalError(
        'Tried to read from a GitPackIndex with no packfile loaded into memory'
      )
    }
    const raw = (await this.pack).slice(start)
```

The code should await the promise before the check:

```javascript
    const pack = await this.pack;
    if (!pack) {
      throw new InternalError(
        'Tried to read from a GitPackIndex with no packfile loaded into memory'
      )
    }
    const raw = pack.slice(start)
```

Also, we can add last error to FS
```javascript
  async read(filepath, options = {}) {
    try {
      let buffer = await this._readFile(filepath, options)
      // Convert plain ArrayBuffers to Buffers
      if (typeof buffer !== 'string') {
        buffer = Buffer.from(buffer)
      }
      return buffer
    } catch (err) {
      this._last_err = err;
      return null
    }
  }
```

with this, we can throw a better error which explains why the pack file was not loaded.

**Comment by mojavelinux** [comment]
Here's the error I see:

```
RangeError [ERR_FS_FILE_TOO_LARGE]: File size (2369120843) is greater than 2 GiB
```

**Comment by mojavelinux** [comment]
The workaround is to configure git so it limits the size of the pack files it creates locally. This is done using the pack.packSizeLimit config setting.

```
git config pack.packSizeLimit 1g
git gc
```

You can find information about this setting here: https://www.git-scm.com/docs/git-pack-objects

**Comment by mojavelinux** [comment]
> The code should await the promise before the check:

Yes, but the message should be adjusted to say that there was an IO error reading the packfile. That's different from the packfile not being loaded.

**Comment by mojavelinux** [comment]
> with this, we can throw a better error which explains why the pack file was not loaded.

Seems reasonable. Or, read could be changed to not suppress the IO error and rather catch it when called.
